genvm/vm.go

```diff
@@ -0,0 +1,307 @@
+		raw:     raw,
+	}
+}
+
+// GetNonce returns expected next nonce for the address.
+func (v *VM) GetNonce(address core.Address) (core.Nonce, error) {
+	account, err := accounts.Latest(v.db, types.Address(address))
+	if err != nil {
+		return core.Nonce{}, err
+	}
+	return core.Nonce{Counter: account.NextNonce()}, nil
+}
+
+// ApplyGenesis saves list of accounts for genesis.
+func (v *VM) ApplyGenesis(genesis []types.Account) error {
+	tx, err := v.db.Tx(context.Background())
+	if err != nil {
+		return err
+	}
+	defer tx.Release()
+	for i := range genesis {
+		account := &genesis[i]
+		v.logger.With().Info("genesis account", log.Inline(account))
+		if err := accounts.Update(tx, account); err != nil {
+			return fmt.Errorf("inserting genesis account %w", err)
+		}
+	}
+	return tx.Commit()
+}
+
+// Apply transactions.
+func (v *VM) Apply(lid types.LayerID, txs []types.RawTx, rewards []types.AnyReward) ([]types.TransactionID, error) {
+	tx, err := v.db.Tx(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Release()
+	var (
+		ss      = core.NewStagedCache(tx)
+		rd      bytes.Reader
+		decoder = scale.NewDecoder(&rd)
+		skipped []types.TransactionID
+		start   = time.Now()
+		fees    uint64
+	)
+	for i := range txs {
+		tx := &txs[i]
+		rd.Reset(tx.Raw)
+		header, ctx, args, err := parse(v.logger, v.registry, ss, tx.ID, decoder)
+		if err != nil {
+			v.logger.With().Warning("skipping transaction. failed to parse", log.Err(err))
+			skipped = append(skipped, tx.ID)
+			continue
+		}
+		v.logger.With().Debug("applying transaction",
+			log.Object("header", header),
+			log.Object("account", &ctx.Account),
+		)
+		if ctx.Account.Balance < ctx.Header.MaxGas*ctx.Header.GasPrice {
+			v.logger.With().Warning("skipping transaction. can't cover gas",
+				log.Object("header", header),
+				log.Object("account", &ctx.Account),
+			)
+			skipped = append(skipped, tx.ID)
+			continue
+		}
+		if ctx.Account.NextNonce() != ctx.Header.Nonce.Counter {
+			v.logger.With().Warning("skipping transaction. failed nonce check",
+				log.Object("header", header),
+				log.Object("account", &ctx.Account),
+			)
+			skipped = append(skipped, tx.ID)
+			continue
+		}
+		if err := ctx.Handler.Exec(ctx, ctx.Header.Method, args); err != nil {
+			v.logger.With().Debug("transaction failed",
+				log.Object("header", header),
+				log.Object("account", &ctx.Account),
+				log.Err(err),
+			)
+			// TODO anything but internal must be recorded in the execution result.
+			// internal errors are propagated upwards, but they are for fatal
+			// unrecovarable errors, such as disk problems
+			if errors.Is(err, core.ErrInternal) {
+				return nil, err
+			}
+		}
+		fee, err := ctx.Apply(ss)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %s", core.ErrInternal, err.Error())
+		}
+		fees += fee
+	}
+
+	// TODO(dshulyak) why do we fail if there are no rewards? can we just burn them?
+	if len(rewards) > 0 {
```

Comment:
why is it invalid? can't there be a consensus on a block without rewards? do we verify it when accepting a block?

Reply:
we don't verify it now, but think we should. 

i don't know if rewards was part of the consensus. but as current code stands, producing a block without rewards is not following the protocol.

Reply:
i opened this issue so that we wont forget https://github.com/spacemeshos/go-spacemesh/issues/3251

but i really doubt that it makes sense to invalidate such block. for example what is the difference between block with empty rewards and the block with rewards given to random non-existing addresses? but lets continue in the issue